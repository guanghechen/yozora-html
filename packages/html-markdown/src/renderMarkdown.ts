import type { Definition, FootnoteDefinition, Root } from '@yozora/ast'
import type { INodeRendererMap } from './rendererMap'
import { createNodesRendererContext, defaultRendererMap } from './rendererMap'
import { renderFootnoteDefinitions } from './renderFootnoteDefinitions'

export function renderMarkdown(
  ast: Root,
  definitionMap: Record<string, Definition>,
  footnoteDefinitionMap: Record<string, FootnoteDefinition>,
  rendererMap: INodeRendererMap = defaultRendererMap,
): string {
  const context = createNodesRendererContext(definitionMap, footnoteDefinitionMap, rendererMap)
  const footnotes: string = renderFootnoteDefinitions(Object.values(footnoteDefinitionMap), context)
  const children: string = context.renderChildren(ast.children)

  /* prettier-ignore */
  return (
    `<section class="yozora-markdown">` +
      `<main>${children}</main>` +
      `<footer>${footnotes}</footer>` +
    `</section>`
  )
}
