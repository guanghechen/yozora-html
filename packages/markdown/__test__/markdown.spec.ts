import type { IRoot } from '@yozora/ast'
import { HtmlType } from '@yozora/ast'
import { calcDefinitionMap, calcFootnoteDefinitionMap } from '@yozora/ast-util'
import fs from 'fs-extra'
import path from 'path'
import type { YastNodeRendererMap } from '../src'
import renderMarkdown, { defaultRendererMap } from '../src'

const renderMap: YastNodeRendererMap = {
  ...defaultRendererMap,
  [HtmlType]: () => '',
}

const resolveFixture = (...p: string[]): string =>
  path.join(__dirname, 'fixtures', ...p)

const loadYozoraAst = (filepath: string): IRoot =>
  fs.readJSONSync(resolveFixture(filepath.replace(/(\.json)?$/, '.json')))

describe('snapshot', function () {
  it('basic', function () {
    const ast0 = loadYozoraAst('basic')
    const { root: ast1, definitionMap } = calcDefinitionMap(ast0)
    const { root, footnoteDefinitionMap } = calcFootnoteDefinitionMap(ast1)
    expect(
      renderMarkdown(root, definitionMap, footnoteDefinitionMap, renderMap),
    ).toMatchSnapshot()
  })
})
