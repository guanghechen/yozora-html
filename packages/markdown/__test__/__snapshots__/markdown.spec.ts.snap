// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot basic 1`] = `
"<div class=\\"yozora-markdown\\"><section><p class=\\"yozora-paragraph\\"><span class=\\"yozora-text\\">A monorepo contains renderers for rendering </span><a class=\\"yozora-link\\" href=\\"https://www.npmjs.com/package/@yozora/ast\\" title=\\"https://www.npmjs.com/package/@yozora/ast\\" target=\\"_blank\\" rel=\\"noopener,noreferrer\\"></a><span class=\\"yozora-text\\"> into HTML strings.
See </span><a class=\\"yozora-link\\" href=\\"https://yozora.guanghechen.com\\" title=\\"https://yozora.guanghechen.com\\" target=\\"_blank\\" rel=\\"noopener,noreferrer\\"><span class=\\"yozora-text\\">https://yozora.guanghechen.com</span></a><span class=\\"yozora-text\\"> for details.</span></p><p class=\\"yozora-paragraph\\"><span class=\\"yozora-text\\">Use </span><a class=\\"yozora-link\\" href=\\"https://github.com/yozorajs/yozora-html/tree/main/packages/markdown#readme\\" title=\\"https://github.com/yozorajs/yozora-html/tree/main/packages/markdown#readme\\" target=\\"_blank\\" rel=\\"noopener,noreferrer\\"></a><span class=\\"yozora-text\\"> to render the </span><a class=\\"yozora-link\\" href=\\"https://www.npmjs.com/package/@yozora/ast#root\\" title=\\"https://www.npmjs.com/package/@yozora/ast#root\\" target=\\"_blank\\" rel=\\"noopener,noreferrer\\"></a><span class=\\"yozora-text\\"> of Yozora ast.</span></p><h2 className=\\"yozora-heading\\"><p className=\\"yozora-heading__content\\"><span class=\\"yozora-text\\">Overview</span></p></h2><h3 className=\\"yozora-heading\\"><p className=\\"yozora-heading__content\\"><span class=\\"yozora-text\\">Markdown components</span></p></h3><h3 className=\\"yozora-heading\\"><p className=\\"yozora-heading__content\\"><span class=\\"yozora-text\\">Other components</span></p></h3></section><footer><div class=\\"yozora-footnote-definition\\"><div class=\\"yozora-footnote-definition__title\\">footnote-definition</div><ul class=\\"yozora-footnote-definition__main\\"><li id=foo-2 class=\\"yozora-footnote-definition\\"><p class=\\"yozora-footnote-definition__title yozora-paragraph\\"><a href=#reference-foo-2>&uarr;</a><span>foo-2</span></p><div class=\\"yozora-footnote-definition__content\\"><p class=\\"yozora-paragraph\\"><span class=\\"yozora-text\\">This is a footnote definition.</span></p></div></li></ul></div></footer></div>"
`;
